### ISSUE TITLE
training set has multi sites problem

### ISSUE THREAD

**Issue by eliauk07yz** [inesws/Normative_AE#2]
Hi
    
    I also have another problem.If the training set has multiple sites, how do these site effects get rid of？I see that the test set has the ‘refbatch’ parameter, if the training set is multiple sites, how to write the test set parameter？

best wishes

**Comment by inesws** [comment]
If the training set has multiple site I suggest:

-Harmonize the training set a) choosing a reference site (M-ComBat application) or b) using the normal combat with global mean and pooled variance 

if a) create a combat_model instance and do combat_model.fit(training_set)
a1) harmonize the test set to the same reference site of the training set by concatenating the test set with the training set and creating a combat object with ref_batch attribute specifying the reference site ID

if b) 
b1) IF test set has the same sites as in the training set DO : combat_model.fit(training_data) -> combat_model.transform(test_data)
b2) IF test set is a unseen NEW SITE : 
b21) assume the already harmonized training set as a "unique" site and give it a new unique ID (eg: site ID = 0 to all the harmonize training set) 
b22) concatenate the harmonize the training set with the new site 
b23) in combat model choose ref_batch to the unique harmonize training set site ID (eg ref_batch = 0) and you can use the combat model to harmonize the new site

**Comment by eliauk07yz** [comment]
Thank you very much for your reply, but I still have some doubts about part a1（ harmonize the test set to the same reference site of the training set），Do I need to  concatenate  the training set and the test set together and use the combat_model.fit(test_set) method？

**Comment by inesws** [comment]
Yes, you concatenate the test set (or any new sample of at least 10-13 observations ) with the training set (I edited my previous comment)

if you check there is an almost similar example in a notebook: Normative_AE/blob/main/Analysis%20Scrips/Confounder_removal_application.ipynb

In summary:
"
# **Use M-ComBat** Variation:


features_info={ 'cortical': { 'id': cortical_columns,
                                          'categorical': ['Gender'],
                                          'continuous':['Age'] },
                         'volumes': { 'id': volumes_columns,
                                           'categorical': ['Gender'],
                                           'continuous':['Age','TIV'] }
                         }

# Create M_combat object
**ref_batch = 0** # This is where you define the ID of your reference site 

M_combat = ComBatHarmonization( cv_method=None, **ref_batch=ref_batch**,
                                                         regression_fit = 0, feat_detail = feat_info,
                                                         feat_of_no_interest = None )

# Concatenate training set with test set

# Concatenate all HC data together # ->HERE concatenate

all_data_feat = pd.concat([training_set, test_set],axis=0).reset_index(drop=True) # FEATURE DATASET

all_data_covars = pd.concat([training_covars, test_covars],axis=0).reset_index(drop=True) # COVARIATES MATRIX

all_data_covars['Batch'] = all_data_covars['Batch'].astype('int') # MAKE SURE site column is int type
HC_covars.rename( columns={"Batch" : "batch"}, in place = True) # RENAME site column 'name' as 'bactch'

(....)

# Create the dictionaries with data + covariates for harmonization
all_data_dict = {'data': all_data_feat , 'covariates': all_data_covars}

# Fit_transform method 
all_data_feat_harmonized = M_combat.fit_transform(all_data_dict)


""

**Comment by eliauk07yz** [comment]
Thank you for your reply. Sorry to bother you again.
1. If I use part a1, does the test set only need to concatenate part of the training set data from the same reference site as the training set, rather than the entire training set data?
2. Also, if the data test set has health and patient, can I put health and patient together and combat_model.fit(test_set)?
3, fit_transform(all_data_dict) This function is not found

**Comment by inesws** [comment]
1. Yes I suppose only need to concatenate the test set with the reference site;
2. If it has healthy and patients I would recommend using the normative approach. If you put it together you need to give the diagnosis labels to the combat model and this might not be available in the test set;
3. It was just an example, it was an error because I defined the dict name as  "data_dict" ;

**Comment by eliauk07yz** [comment]
Thank you for your reply,  but i can not find this method named fit_transform(),there are only fit() and transform(),can i use those function()?
